upgrade/v122to130/upgrade.go

```diff
@@ -0,0 +1,89 @@
+package v122to130
+
+import (
+	"context"
+	"github.com/sirupsen/logrus"
+
+	"github.com/pkg/errors"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta1"
+	lhclientset "github.com/longhorn/longhorn-manager/k8s/pkg/client/clientset/versioned"
+)
+
+const (
+	upgradeLogPrefix = "upgrade from v1.2.2 to v1.3.0: "
+)
+
+func UpgradeCRs(namespace string, lhClient *lhclientset.Clientset) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, upgradeLogPrefix+"UpgradeCRs failed")
+	}()
+	if err := upgradeEngines(namespace, lhClient); err != nil {
+		return err
+	}
+	return nil
+}
+
+func upgradeEngines(namespace string, lhClient *lhclientset.Clientset) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "upgrade engines failed")
+	}()
+
+	engines, err := lhClient.LonghornV1beta1().Engines(namespace).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	volumeEngineMap := map[string][]*v1beta1.Engine{}
+	for i := range engines.Items {
+		e := &engines.Items[i]
+		if e.Spec.VolumeName == "" {
+			// Cannot do anything in the upgrade path if this case happens.
+			continue
+		}
+		volumeEngineMap[e.Spec.VolumeName] = append(volumeEngineMap[e.Spec.VolumeName], e)
+	}
+
+	for volumeName, engineList := range volumeEngineMap {
+		skip := false
+		for _, e := range engineList {
+			if e.Spec.Active {
+				skip = true
+				break
+			}
+		}
+		if skip {
+			continue
+		}
+
+		var currentEngine *v1beta1.Engine
+		if len(engineList) == 1 {
+			currentEngine = engineList[0]
+		} else {
+			v, err := lhClient.LonghornV1beta1().Volumes(namespace).Get(context.TODO(), volumeName, metav1.GetOptions{})
+			if err != nil {
+				return err
+			}
+			for i := range engineList {
+				if (v.Spec.NodeID != "" && v.Spec.NodeID == engineList[i].Spec.NodeID) ||
+					(v.Status.CurrentNodeID != "" && v.Status.CurrentNodeID == engineList[i].Spec.NodeID) ||
+					(v.Status.PendingNodeID != "" && v.Status.PendingNodeID == engineList[i].Spec.NodeID) {
```

Comment:
Generally, I am thinking we better organize this kind of reusable condition check in some util libraries. Not for now, but it's good to think of it.